src/services/document-service.js

```diff
@@ -248,135 +360,89 @@ class DocumentService {
-      throw new Error(`Revision ${revisionId} cannot be restored, it is the latest revision`);
-    }
+  async copyDocumentRevisions({ revisions, ancestorId, origin, originUrl }) {
+    let lock;
+    const newDocumentRevisions = [];
+    const documentKey = revisions[0].key;
 
-    const doc = {
-      title: revisionToRestore.title,
-      slug: revisionToRestore.slug,
-      namespace: revisionToRestore.namespace,
-      language: revisionToRestore.language,
-      sections: cloneDeep(revisionToRestore.sections),
-      appendTo: {
-        key: documentKey,
-        ancestorId: latestRevision._id
-      },
-      tags: revisionToRestore.tags
-    };
+    try {
+      lock = await this.documentLockStore.takeLock(documentKey);
 
-    await this.createDocumentRevision({ doc, user, restoredFrom: revisionToRestore._id });
+      await this.transactionRunner.run(async session => {
 
-    return this.getAllDocumentRevisionsByKey(documentKey);
-  }
+        const existingDocumentRevisions = await this._getAllDocumentRevisionsByKey(documentKey, session);
+        const latestExistingRevision = existingDocumentRevisions[existingDocumentRevisions.length - 1];
 
-  async hardDeleteSection({ documentKey, sectionKey, sectionRevision, reason, deleteAllRevisions, user }) {
-    if (!user || !user._id) {
-      throw new Error('No user specified');
-    }
+        if (!ancestorId && latestExistingRevision) {
+          throw new Error(`Found unexpected existing revisions for document '${documentKey}'`);
+        }
 
-    let lock;
-    const now = new Date();
-    const userId = user._id;
+        if (ancestorId && latestExistingRevision?._id !== ancestorId) {
+          throw new Error(`Import of document '${documentKey}' expected to find revision '${ancestorId}' as the latest revision but found revision '${latestExistingRevision?._id}'`);
+        }
 
-    try {
+        for (const revision of revisions) {
 
-      logger.info(`Hard deleting sections with section key ${sectionKey} in documents with key ${documentKey}`);
+          const data = { ...revision, origin, originUrl };
 
-      lock = await this.documentLockStore.takeLock(documentKey);
+          // eslint-disable-next-line no-await-in-loop
+          const order = await this.documentOrderStore.getNextOrder();
 
-      const allRevisions = await this.getAllDocumentRevisionsByKey(documentKey);
+          const newDocumentRevision = this._buildDocumentRevision({
+            data,
+            revisionId: revision._id,
+            documentKey,
+            userId: revision.createdBy,
+            order,
+            restoredFrom: revision.restoredFrom,
+            sections: cloneDeep(revision.sections)
+          });
 
-      const revisionsToUpdate = [];
+          logger.info(`Saving new document revision with id ${newDocumentRevision._id}`);
+          // eslint-disable-next-line no-await-in-loop
+          await this.documentRevisionStore.save(newDocumentRevision, { session });
```

Comment:
The only improvement I can still see here, would be to not save each revision within the loop, but to call `insertMany` after the loop is done (and the `newDocumentRevisions` array is filled), but I don't think it will make it that much slower, it's not a performance critical path anyway, so I'll approve regardless for what you decide.

Reply:
Yes we said we wanted to do that actually. I'll want to do that 👍

Reply:
Changed.